azkaban-execserver/src/main/java/azkaban/execapp/event/JobCallbackManager.java

```diff
@@ -0,0 +1,267 @@
+
+  private static final Logger logger = Logger
+      .getLogger(JobCallbackManager.class);
+
+  private static boolean isInitialized = false;
+  private static JobCallbackManager instance;
+
+  private static int maxNumCallBack = 3;
+
+  private JmxJobCallbackMBean callbackMbean;
+  private String azkabanHostName;
+  private SimpleDateFormat gmtDateFormatter;
+
+  private static final JobCallbackStatusEnum[] ON_COMPLETION_JOB_CALLBACK_STATUS =
+      { SUCCESS, FAILURE, COMPLETED };
+
+  public static void initialize(Props props) {
+    logger.info("Initializing");
+    instance = new JobCallbackManager(props);
+
+    isInitialized = true;
+  }
+
+  public static boolean isInitialized() {
+    return isInitialized;
+  }
+
+  public static JobCallbackManager getInstance() {
+    if (!isInitialized) {
+      throw new IllegalStateException(JobCallbackManager.class.getName()
+          + " has not been initialized");
+    }
+    return instance;
+  }
+
+  private JobCallbackManager(Props props) {
+    maxNumCallBack = props.getInt("jobcallback.max_count", maxNumCallBack);
+
+    // initialize the request maker
+    JobCallbackRequestMaker.initialize(props);
+
+    callbackMbean =
+        new JmxJobCallback(JobCallbackRequestMaker.getInstance()
+            .getJobcallbackMetrics());
+
+    azkabanHostName = getAzkabanHostName(props);
+
+    gmtDateFormatter = new SimpleDateFormat("EEE, dd MMM yyyy HH:mm:ss z");
+    gmtDateFormatter.setTimeZone(TimeZone.getTimeZone("GMT"));
+
+    logger.info("Initialization completed " + getClass().getName());
+    logger.info("azkabanHostName " + azkabanHostName);
+  }
+
+  public JmxJobCallbackMBean getJmxJobCallbackMBean() {
+    return callbackMbean;
+  }
+
+  @Override
+  public void handleEvent(Event event) {
+    if (!isInitialized) {
+      return;
+    }
+
+    if (event.getRunner() instanceof JobRunner) {
+      try {
+        if (event.getType() == Event.Type.JOB_STARTED) {
+          processJobCallOnStart(event);
+        } else if (event.getType() == Event.Type.JOB_FINISHED) {
+          processJobCallOnFinish(event);
+        }
+      } catch (Throwable e) {
+        // Use job runner logger so user can see the issue in their job log
+        JobRunner jobRunner = (JobRunner) event.getRunner();
+        jobRunner.getLogger().error(
+            "Encountered error while hanlding job callback event", e);
+      }
+    } else {
+      logger.warn("((( Got an unsupported runner: "
+          + event.getRunner().getClass().getName() + " )))");
+    }
+
+  }
+
+  private void processJobCallOnFinish(Event event) {
+    JobRunner jobRunner = (JobRunner) event.getRunner();
+
+    if (!JobCallbackUtil.isThereJobCallbackProperty(jobRunner.getProps(),
+        ON_COMPLETION_JOB_CALLBACK_STATUS)) {
+      return;
+    }
+
+    // don't want to waste time resolving properties if there are no
+    // callback properties to parse
+    Props props = PropsUtils.resolveProps(jobRunner.getProps());
```

Comment:
Do you need to resolve props ? Azkaban should have done it before starting execution ?